syntax = "proto3";

package library;

// Book message
message Book {
    int32 id = 1;
    string title = 2;
    string author = 3;
    bool is_borrowed = 4;
    int32 current_member_id = 5;
    string current_member_name = 6; // Member name if borrowed
}

// Member message
message Member {
    int32 id = 1;
    string name = 2;
    string email = 3;
}

// Ledger entry message
message LedgerEntry {
    int32 id = 1;
    int32 book_id = 2;
    int32 member_id = 3;
    string action_type = 4;
    string log_date = 5;
    string due_date_snapshot = 6;
}

// Create Book Request/Response
message CreateBookRequest {
    string title = 1;
    string author = 2;
}

message CreateBookResponse {
    Book book = 1;
    string message = 2;
}

// Update Book Request/Response
message UpdateBookRequest {
    int32 id = 1;
    string title = 2;
    string author = 3;
}

message UpdateBookResponse {
    Book book = 1;
    string message = 2;
}

// List Books Request/Response
message ListBooksRequest {
}

message ListBooksResponse {
    repeated Book books = 1;
}

// Create Member Request/Response
message CreateMemberRequest {
    string name = 1;
    string email = 2;
}

message CreateMemberResponse {
    Member member = 1;
    string message = 2;
}

// Update Member Request/Response
message UpdateMemberRequest {
    int32 id = 1;
    string name = 2;
    string email = 3;
}

message UpdateMemberResponse {
    Member member = 1;
    string message = 2;
}

// Borrow Book Request/Response
message BorrowBookRequest {
    int32 book_id = 1;
    int32 member_id = 2;
}

message BorrowBookResponse {
    bool success = 1;
    string message = 2;
    LedgerEntry ledger_entry = 3;
}

// Return Book Request/Response
message ReturnBookRequest {
    int32 book_id = 1;
    int32 member_id = 2;
}

message ReturnBookResponse {
    bool success = 1;
    string message = 2;
    LedgerEntry ledger_entry = 3;
}

// List Borrowed Books Request/Response
message ListBorrowedBooksRequest {
    int32 member_id = 1;
}

message ListBorrowedBooksResponse {
    repeated Book books = 1;
}

// Search Books Request/Response
message SearchBooksRequest {
    string query = 1; // Search by title or author
}

message SearchBooksResponse {
    repeated Book books = 1;
}

// List Members Request/Response
message ListMembersRequest {
}

message ListMembersResponse {
    repeated Member members = 1;
}

// Search Members Request/Response
message SearchMembersRequest {
    string query = 1; // Search by name or email
}

message SearchMembersResponse {
    repeated Member members = 1;
}

// Library Service
service LibraryService {
    rpc CreateBook(CreateBookRequest) returns (CreateBookResponse);
    rpc UpdateBook(UpdateBookRequest) returns (UpdateBookResponse);
    rpc ListBooks(ListBooksRequest) returns (ListBooksResponse);
    rpc SearchBooks(SearchBooksRequest) returns (SearchBooksResponse);
    
    rpc CreateMember(CreateMemberRequest) returns (CreateMemberResponse);
    rpc UpdateMember(UpdateMemberRequest) returns (UpdateMemberResponse);
    rpc ListMembers(ListMembersRequest) returns (ListMembersResponse);
    rpc SearchMembers(SearchMembersRequest) returns (SearchMembersResponse);
    
    rpc BorrowBook(BorrowBookRequest) returns (BorrowBookResponse);
    rpc ReturnBook(ReturnBookRequest) returns (ReturnBookResponse);
    rpc ListBorrowedBooks(ListBorrowedBooksRequest) returns (ListBorrowedBooksResponse);
}

